utils_nlp/common/pytorch_utils.py

```diff
@@ -32,59 +28,107 @@ def get_device(
+        Module, DataParallel, DistributedDataParallel: A PyTorch Module or
+            a DataParallel/DistributedDataParallel wrapper (when multiple gpus are used).
     """
-    if isinstance(model, nn.DataParallel):
-        model = model.module
-
     if not isinstance(device, torch.device):
         raise ValueError("device must be of type torch.device.")
 
-    if device.type == "cuda":
-        model.to(device)  # inplace
-        if num_gpus == 0:
-            raise ValueError("num_gpus must be non-zero when device.type is 'cuda'")
-        elif num_gpus == 1:
-            return model
-        else:
-            # parallelize
+    # unwrap model
+    if isinstance(model, torch.nn.DataParallel):
+        model = model.module
+    # wrap in DataParallel or DistributedDataParallel
+    if local_rank != -1:
+        self.model = torch.nn.parallel.DistributedDataParallel(
+            self.model, device_ids=[local_rank], output_device=local_rank, find_unused_parameters=True,
+        )
+    else:
+        if device.type == "cuda":
+            if num_gpus is not None:
+                if num_gpus < 1:
+                    raise ValueError("num_gpus must be at least 1 or None")
             num_cuda_devices = torch.cuda.device_count()
             if num_cuda_devices < 1:
                 raise Exception("CUDA devices are not available.")
-            elif num_cuda_devices < 2:
-                print("Warning: Only 1 CUDA device is available. Data parallelism is not possible.")
-                return model
-            else:
-                if num_gpus is None:
-                    # use all available devices
-                    return nn.DataParallel(model, device_ids=None)
-                elif num_gpus > num_cuda_devices:
-                    print(
-                        "Warning: Only {0} devices are available. "
-                        "Setting the number of gpus to {0}".format(num_cuda_devices)
-                    )
-                    return nn.DataParallel(model, device_ids=None)
-                else:
-                    return nn.DataParallel(model, device_ids=list(range(num_gpus)))
-    elif device.type == "cpu":
-        if num_gpus != 0 and num_gpus is not None:
-            warnings.warn("Device type is 'cpu'. num_gpus is ignored.")
-        return model.to(device)
+            if gpu_ids is None:
+                num_gpus = num_cuda_devices if num_gpus is None else min(num_gpus, num_cuda_devices)
+                gpu_ids = list(range(num_gpus))
+            if len(gpu_ids) > 1:
+                model = torch.nn.DataParallel(model, device_ids=gpu_ids)
+    # move to device
+    return model.to(device)
+
 
+def dataloader_from_dataset(ds, batch_size=32, num_gpus=None, shuffle=False, distributed=False):
+    """Creates a PyTorch DataLoader given a Dataset object.
+
+    Args:
+        ds (torch.utils.data.DataSet): A PyTorch dataset.
+        batch_size (int, optional): Batch size. Defaults to 32.
```

Comment:
Can explain batch_size is batch_size per gpu in the docstring? Would be even better to name the argument batch_size_per_gpu to make it more explicit, but I'll leave it up to you.